llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+
+class PartialInterpreter : public llvm::Interpreter {
+	friend FunctionData;
+	friend ModuleData;
+	const llvm::BasicBlock* fromBB{nullptr};	
+	enum BitMask : uint32_t
+	{
+		NONE = 0,
+		ALIGNED2 = 0x00000001,
+		ALIGNED4 = 0x00000003,
+		ALIGNED8 = 0x00000007,
+		ALL = 0xffffffff,
+	};
+	std::vector<std::pair<const llvm::Value*, const llvm::Value*> > incomingsVal;
+	std::vector<const llvm::PHINode*> notComputedPhis;
+	std::vector<std::pair<long long,long long>> validLoadIntervals;
+public:
+	GenericValue getOperandValueExplicit(const Value* V, ExecutionContext& EC)
+	{
+		return Interpreter::getOperandValue(const_cast<Value*>(V), EC);
+	}
+	GenericValue getOperandValue(const Value* V)
+	{
+		return getOperandValueExplicit(V, getLastStack());
+	}
+	void assignToMaps(const Value* V, const BitMask bitMask)
+	{
+		stronglyKnownBits[V] = bitMask;
+	}
+	void assignToMaps(const Value* V, const BitMask bitMask, GenericValue GV)
+	{
+		assignToMaps(V, bitMask);
+		getLastStack().Values[const_cast<Value*>(V)] = GV;
+	}
+	void assignToMaps(const Value* V, const Value* toAssign)
+	{
+		assignToMaps(V, getBitMask(toAssign), getOperandValue(toAssign)); 
+	}
+	void assignToMapsIfComputed(const Value* V, const Value* toAssign)
+	{
+		if (isValueComputed(toAssign))
+		{
+			assignToMaps(V, getBitMask(toAssign));
+		}
+	}
+	void removeFromMaps(const Value* V)
+	{
+		stronglyKnownBits.erase(V);
+		getLastStack().Values.erase(const_cast<Value*>(V));
+	}
+	
+	llvm::DenseMap<const llvm::Value*, BitMask> stronglyKnownBits;
+	BitMask computeStronglyKnownBits(const llvm::Instruction& I)
+	{
+		BitMask min = BitMask::ALL;
+
+		for (auto& op : I.operands())
+		{
+			min = BitMask(min & getBitMask(op));
+		}
+
+		if (min == BitMask::ALL)
+		{
+			//All operands fully known
+			return BitMask::ALL;
+		}
+
+		switch (I.getOpcode())
+		{
+			case Instruction::Load:
+			{
+				return BitMask::ALL;
+			}
+			case Instruction::Select:
+				// TODO(carlo): Select can also be specialised further
+			case Instruction::Add:
+			case Instruction::Sub:
+			case Instruction::Mul:
+			case Instruction::PtrToInt:
+			case Instruction::BitCast:
+			case Instruction::GetElementPtr:
+			{
+				//Those all work modulo N
+				return min;
+			}
+			case Instruction::And:
+			{
+				//If either is Constant && less than equal min -> ALL
+				if (const ConstantInt* CI = dyn_cast<ConstantInt>(I.getOperand(0)))
+				{
+					if (CI->getZExtValue() <= min)
+						return BitMask::ALL;
+				}
+				if (const ConstantInt* CI = dyn_cast<ConstantInt>(I.getOperand(1)))
+				{
+					if (CI->getZExtValue() <= min)
+						return BitMask::ALL;
+				}
+
+				[[clang::fallthrough]];
+			}
+			case Instruction::Or:
+			case Instruction::Xor:
+			{
+				//Those all work modulo 2^k
+				return min;
+			}
+			default:
+			{
+				//Some operands NOT fully known, means in general no information is conserved
+				return BitMask::NONE;
+			}
+		}
+	}
+	void setIncomingBB(const llvm::BasicBlock* from)
```

Comment:
This setter/getter are only used in 1 place, and also names just like an internal LLVM function. Let's just use the variable.